lib/si-service/src/shutdown.rs

```diff
@@ -42,36 +42,44 @@ impl ShutdownError {
 ///
 /// This function sets up a signal handler for both `SIGINT` (i.e. `Ctrl+c`) and `SIGTERM` so usage
 /// of this function with other code intercepting these signals is *highly* discouraged.
-pub async fn graceful<Fut, E>(
-    tracker: TaskTracker,
-    token: CancellationToken,
+pub async fn graceful<Fut, E, I>(
+    trackers_with_tokens: I,
     telemetry_guard: Option<Fut>,
     shutdown_timeout: Option<Duration>,
 ) -> Result<(), ShutdownError>
 where
     Fut: Future<Output = Result<(), E>>,
     E: error::Error + Send + Sync + 'static,
+    I: IntoIterator<Item = (TaskTracker, CancellationToken)>,
 {
     let mut sig_int = unix::signal(SignalKind::interrupt()).map_err(ShutdownError::Signal)?;
     let mut sig_term = unix::signal(SignalKind::terminate()).map_err(ShutdownError::Signal)?;
 
     tokio::select! {
         _ = sig_int.recv() => {
             info!("received SIGINT, performing graceful shutdown");
-            tracker.close();
-            token.cancel();
         }
         _ = sig_term.recv() => {
             info!("received SIGTERM, performing graceful shutdown");
+        }
+    }
+
+    // Create a future for draining each task group
+    let drain_future = async {
+        let mut count = 0;
+        for (tracker, token) in trackers_with_tokens {
+            count += 1;
+            info!("performing graceful shutdown for task group {count}");
             tracker.close();
             token.cancel();
+            tracker.wait().await;
         }
-    }
+    };
```

Comment:
The PR and commit descriptions get into this further, but using a manger struct could result in better tracing when draining (e.g. "we are now draining the telemetry group").

For now, let's use the raw types as we are refactoring frequently here, at present.